sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala

```diff
@@ -1535,6 +1539,211 @@ class ParquetIOSuite extends QueryTest with ParquetTest with SharedSparkSession
+    withTempPaths(2) { paths =>
+      paths.foreach(_.delete())
+      val df = (0 to 10000).toDF("id")
+      val rowIndexColName = RowIndexGenerator.ROW_INDEX_COLUMN_NAME
+
+      val pathWithNoRowIdx = paths.head.getAbsolutePath
+      val pathWithRowIdx = paths(1).getAbsolutePath
+
+      // no row index in schema.
+      df.repartition(1).write.format(dataSourceName).save(pathWithNoRowIdx)
+      val noRowIdxDF = spark.read.format(dataSourceName).load(pathWithNoRowIdx)
+      assert(!noRowIdxDF.columns.contains(rowIndexColName))
+
+      // With row index in schema.
+      val schemaWithRowIdx = df.schema.add(rowIndexColName, LongType, nullable = true)
+
+      df.repartition(1)
+        .write
+        .format(dataSourceName)
+        .option("parquet.block.size", 128) // Force the data to be split into multiple row groups.
+        .save(pathWithRowIdx)
+
+      val dfWithOnePartition = spark.read
+        .format(dataSourceName)
+        .schema(schemaWithRowIdx)
+        .load(pathWithRowIdx)
+      assert(dfWithOnePartition.filter(s"$rowIndexColName != id").count() == 0)
+    }
+  }
+
+  case class RowGroupStat(
+                           idx: Int,
+                           rowCnt: Long,
+                           offset: Long,
+                           byteSize: Long,
+                           isLast: Boolean
+                         )
+
+  def checkRowGroups(path: String, split: Option[(Long, Long)] = None): Seq[RowGroupStat] = {
+    val filter = split match {
+      case Some((start, end)) => ParquetMetadataConverter.range(start, end)
+      case None => NO_FILTER
+    }
+    val footer = ParquetFileReader.readFooter(
+      spark.sessionState.newHadoopConf(),
+      new Path(path),
+      filter)
+
+    /*
+     ParquetReadOptions options = HadoopReadOptions
+      .builder(configuration, file)
+      .withRange(split.getStart(), split.getStart() + split.getLength())
+      .build();
+     */
+    val blocks = footer.getBlocks.asScala
+    blocks.zipWithIndex.map { case (rowGroup, rowGroupIdx) =>
+      val rowCnt = rowGroup.getRowCount()
+      // Number of pages is convoluted to get.
+      val byteSize = rowGroup.getCompressedSize()
+      val offset = rowGroup.getRowIndexOffset()
+      val isLast = (rowGroupIdx + 1 == blocks.size)
+      RowGroupStat(rowGroupIdx, rowCnt, offset, byteSize, isLast)
+    }
+  }
+
+  def listParquetFiles(f: File): Seq[File] = {
+    f.listFiles().filter(_.isFile).filter(_.getName().endsWith("parquet"))
+  }
+
+  def dumpParquetFileInfo(dir: File): Unit = {
+    val files = listParquetFiles(dir)
+    for (f <- files) {
+      println(f.getAbsolutePath)
+      println("full:")
+      for (s <- checkRowGroups(f.getAbsolutePath)) {
+        println(s.toString)
+      }
+      println("filtered:")
+      val length = f.length()
+      for (s <- checkRowGroups(f.getAbsolutePath, Some((length/2, length)))) {
+        println(s.toString)
+      }
+    }
+  }
```

Comment:
We read the footer two times: once without a push-down filter, once with push-down filter saying we only want to read the second part of the file.